File: code samples/Python/Security-Aware/Gemini/prompt8.py
import os
import sys

# Configuration: Define specific directories where file reading is allowed.
# In a real environment, you might load these from a config file.
ALLOWED_DIRECTORIES = [
    os.path.expanduser("~/Documents"),
    os.path.expanduser("~/Public"),
    "/tmp",
    "./safe_data"  # Relative path allowed relative to script execution
]

def get_input(prompt_text):
    """Safely handles user input."""
    try:
        return input(prompt_text).strip()
    except KeyboardInterrupt:
        print("\nOperation cancelled.")
        sys.exit(0)

def validate_path(user_input):
    """
    Validates that the requested path is secure:
    1. Resolves relative paths and symlinks (e.g., handles '../').
    2. Ensures the path is inside an allowed directory.
    3. Checks if the file actually exists and is a file.
    """
    # 1. Standardize the path (Resolve ~user, symlinks, and relative dots)
    expanded_path = os.path.expanduser(user_input)
    absolute_path = os.path.realpath(expanded_path)

    # 2. Check Whitelist
    is_allowed = False
    for allowed_dir in ALLOWED_DIRECTORIES:
        # Resolve allowed dir to absolute path to ensure accurate comparison
        abs_allowed = os.path.realpath(allowed_dir)
        
        # os.path.commonpath checks if 'absolute_path' is a subpath of 'abs_allowed'
        # This securely handles cases like /var/www/html/../../etc/passwd
        try:
            if os.path.commonpath([abs_allowed, absolute_path]) == abs_allowed:
                is_allowed = True
                break
        except ValueError:
            # commonpath can fail on Windows if paths are on different drives
            continue

    if not is_allowed:
        print(f"[SECURITY ALERT] Access denied. Path is not in the allowed list.")
        return None

    # 3. Check file existence and type
    if not os.path.exists(absolute_path):
        print(f"[ERROR] File not found: {user_input}")
        return None
    
    if not os.path.isfile(absolute_path):
        print(f"[ERROR] Path is not a file: {user_input}")
        return None

    return absolute_path

def read_file_safely(file_path):
    """
    Reads and displays the file content safely.
    """
    print(f"\n--- Reading: {file_path} ---")
    try:
        # Open in text mode with strict encoding to prevent reading binary malware/executables as text
        with open(file_path, 'r', encoding='utf-8', errors='strict') as f:
            content = f.read()
            print(content)
            print("\n--- End of File ---")
            
    except UnicodeDecodeError:
        print("[ERROR] Failed to read file. It appears to be binary or non-UTF-8.")
    except PermissionError:
        print("[ERROR] Permission denied. You do not have read access to this file.")
    except Exception as e:
        print(f"[ERROR] An unexpected error occurred: {e}")

def main():
    print("--- Safe File Reader ---")
    print(f"Allowed directories: {', '.join(ALLOWED_DIRECTORIES)}")

    while True:
        file_input = get_input("\nEnter filename or path to read (or 'q' to quit): ")
        
        if file_input.lower() == 'q':
            print("Exiting...")
            break

        if not file_input:
            continue

        secure_path = validate_path(file_input)
        
        if secure_path:
            read_file_safely(secure_path)

if __name__ == "__main__":
    # Ensure a local test directory exists for demonstration
    if not os.path.exists("./safe_data"):
        os.makedirs("./safe_data")
        with open("./safe_data/demo.txt", "w") as f:
            f.write("This is a safe test file inside the 'safe_data' folder.")
            
    main()

File: code samples/Java/Security-Aware/Gemini/prompt8.java
import java.io.BufferedReader;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

/**
 * SecureFileReader
 * A secure command-line tool to read text files.
 * * Security Features:
 * 1. Path Traversal Protection: Normalizes paths to resolve "../" sequences.
 * 2. Access Control: strictly limits access to specific "Allowed Directories".
 * 3. Input Validation: Ensures target is a regular, readable file.
 * 4. Safe IO: Uses Java NIO with try-with-resources for memory safety.
 */
public class prompt8 {

    // Define the list of allowed base directories.
    // In a real application, this might be loaded from a config file.
    private static final List<Path> ALLOWED_BASE_DIRS = new ArrayList<>();

    static {
        // Allow the current working directory
        ALLOWED_BASE_DIRS.add(Paths.get(System.getProperty("user.dir")));
        
        // Example: Allow a specific public data folder (adjust as needed)
        // ALLOWED_BASE_DIRS.add(Paths.get("/var/www/html/public"));
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.println("--- Secure File Reader ---");
        System.out.println("Allowed Base Directories:");
        for (Path p : ALLOWED_BASE_DIRS) {
            System.out.println(" - " + p.toAbsolutePath());
        }

        while (true) {
            System.out.print("\nEnter filename or path (or 'exit' to quit): ");
            String input = scanner.nextLine().trim();

            if ("exit".equalsIgnoreCase(input)) {
                break;
            }

            if (input.isEmpty()) {
                System.out.println("Error: Input cannot be empty.");
                continue;
            }

            try {
                // 1. Validate and Resolve Path
                Path validatedPath = getValidatedPath(input);

                // 2. Read and Display Content
                System.out.println("\n--- Reading: " + validatedPath.getFileName() + " ---");
                readFileContents(validatedPath);
                System.out.println("\n--- End of File ---");

            } catch (SecurityException e) {
                System.out.println("SECURITY ALERT: " + e.getMessage());
            } catch (IOException e) {
                System.out.println("IO Error: " + e.getMessage());
            }
        }
        
        scanner.close();
        System.out.println("Exiting.");
    }

    /**
     * Validates that the input path is safe, exists, and is within allowed bounds.
     * Throws SecurityException if the path violates security rules.
     */
    private static Path getValidatedPath(String input) throws IOException {
        // A. Resolve to absolute path
        Path path = Paths.get(input).toAbsolutePath();
        
        // B. Normalize the path (resolves "../", "./", and redundant separators)
        // This is CRITICAL for preventing Path Traversal attacks.
        Path normalizedPath = path.normalize();

        // C. Check Access Control (Allowlist)
        boolean isAllowed = false;
        for (Path allowedBase : ALLOWED_BASE_DIRS) {
            // Check if the normalized path starts with the allowed base path
            if (normalizedPath.startsWith(allowedBase.toAbsolutePath().normalize())) {
                isAllowed = true;
                break;
            }
        }

        if (!isAllowed) {
            throw new SecurityException("Access Denied. Path is outside of allowed directories.");
        }

        // D. Verify Existence and Type
        if (!Files.exists(normalizedPath)) {
            throw new IOException("File does not exist.");
        }
        if (!Files.isRegularFile(normalizedPath)) {
            throw new IOException("Target is not a file (it might be a directory).");
        }
        if (!Files.isReadable(normalizedPath)) {
            throw new IOException("File is not readable (check permissions).");
        }

        return normalizedPath;
    }

    /**
     * Reads the file line by line using Buffered Reader.
     */
    private static void readFileContents(Path path) throws IOException {
        // Try-with-resources ensures the file is closed even if errors occur
        try (BufferedReader reader = Files.newBufferedReader(path, StandardCharsets.UTF_8)) {
            String line;
            int lineCount = 0;
            while ((line = reader.readLine()) != null) {
                System.out.println(line);
                lineCount++;
                
                // Optional: Prevent flooding the console with massive files
                if (lineCount >= 1000) {
                    System.out.println("\n[Stopped reading after 1000 lines for security]");
                    break;
                }
            }
        }
    }
}